### PR TITLE
DR-471 custodians and stewards read access on studies

### PR DESCRIPTION
**Pull request by thathert**

### CODE REVIEW

File: src/main/java/bio/terra/service/SamClientService.java [hunk 1]
```diff
@@ -366,12 +365,10 @@ public void createResourceCorrectCall(
                 "Missing the required parameter 'resourceCreate' when calling createResource(Async)");
         }
 
-        Object localVarPostBody = resourceCreate;
-
         // create path and map variables
         String localVarPath = "/api/resources/v1/{resourceTypeName}"
             .replaceAll("\\{" + "resourceTypeName" + "\\}",
-                localVarApiClient.escapeString(resourceTypeName.toString()));
+                localVarApiClient.escapeString(resourceTypeName));
```
Comment (ddietterich) [inline src/main/java/bio/terra/service/SamClientService.java]: What is this doing?  It looks like it is equivalent to:
```
String localVarPath = "/api/resources/v1/" + localVarApiClient.escapeString(resourceTypeName));
```

File: src/main/java/bio/terra/flight/study/create/CreateStudyAuthzResource.java [hunk 2]
```diff
@@ -13,26 +16,37 @@
...
     public StepResult doStep(FlightContext context) {
         FlightMap inputParameters = context.getInputParameters();
         AuthenticatedUserRequest userReq = inputParameters.get(
             JobMapKeys.USER_INFO.getKeyName(), AuthenticatedUserRequest.class);
         FlightMap workingMap = context.getWorkingMap();
         UUID studyId = workingMap.get("studyId", UUID.class);
+        Study study = studyService.retrieve(studyId);
         try {
-            sam.createStudyResource(userReq, studyId);
+            List<String> policyEmails = sam.createStudyResource(userReq, studyId);
```
Comment (ddietterich) [inline src/main/java/bio/terra/flight/study/create/CreateStudyAuthzResource.java:47]: What is a policy Email?
Maybe a comment here on who is getting access to what would be useful.

File: src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java [hunk 3]
```diff
@@ -248,10 +248,19 @@ public void createDataset(bio.terra.metadata.Dataset dataset, List<String> rowId
...
         BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
         bigQueryProject.addDatasetAcls(dataset.getName(),
-            Collections.singletonList(Acl.of(new Acl.Group(readersEmail), Acl.Role.READER)));
+            Collections.singletonList(Acl.of(new Acl.Group(readerPolicyGroupEmail), Acl.Role.READER)));
+    }
+
+    @Override
+    public void grantReadAccessToStudy(Study study, List<String> policyGroupEmails) {
+        BigQueryProject bigQueryProject = bigQueryProjectForStudy(study);
+        bigQueryProject.addDatasetAcls(prefixName(study.getName()), policyGroupEmails
```
Comment (ddietterich) [inline src/main/java/bio/terra/pdao/bigquery/BigQueryPdao.java]: Nit: I'm not a fan of embedded streams like that. Perhaps build the collection first and then make the add call?

File: src/main/java/bio/terra/service/SamClientService.java [hunk 4]
```diff
@@ -245,6 +231,18 @@ public void createStudyResource(AuthenticatedUserRequest userReq, UUID studyId)
...
+
+        // we'll want all of these roles to have read access to the underlying data, so we sync and return the emails
+        // for the policies that get created by SAM
+        ArrayList<String> rolePolicies = new ArrayList<>();
+        for (DataRepoRole role : Arrays.asList(DataRepoRole.STEWARD, DataRepoRole.CUSTODIAN, DataRepoRole.INGESTER)) {
+            Map<String, List<Object>> results = samGoogleApi(userReq.getToken()).syncPolicy(
+                ResourceType.STUDY.toString(),
+                studyId.toString(),
+                role.toString());
+            rolePolicies.add(results.keySet().iterator().next());
```
Comment (ddietterich) [inline src/main/java/bio/terra/service/SamClientService.java]: Maybe explain what this is doing? Why is getting the `next` the right key to get? Are you guaranteed there is only one?  Some comments here would help.

File: src/main/java/bio/terra/service/StudyService.java [hunk 5]
```diff
@@ -62,10 +62,13 @@ public StudySummaryModel createStudy(StudyRequestModel studyRequest, Authenticat
         return getResponse(flightId, StudySummaryModel.class);
     }
 
-    public StudyModel retrieve(UUID id) {
+    public StudyModel retrieveModel(UUID id) {
```
Comment (ddietterich) [inline src/main/java/bio/terra/service/StudyService.java]: Dup of what is in DR-466. Do you need to create that collision?

**Review by ddietterich — APPROVED** [review]
What I understand LGTM. My comments are almost all "please add comments so I understand what this code is trying to do."

File: src/main/java/bio/terra/service/SamClientService.java (hunk 4 above)
Comment (thathert) [reply src/main/java/bio/terra/service/SamClientService.java]: Yep, good point. This is copying another piece of code that is also not documented -- probably worth factoring out into a separate method with docs.

File: src/main/java/bio/terra/service/StudyService.java (hunk 5 above)
Comment (thathert) [reply src/main/java/bio/terra/service/StudyService.java]: I needed to, but will rebase now that 466 is merged in.

File: src/main/java/bio/terra/service/SamClientService.java (hunk 1 above)
Comment (thathert) [reply src/main/java/bio/terra/service/SamClientService.java]: Yeah.. I think it should be what you wrote here. No need to do fancy string interpolation here.

**Review by RoriCremer — APPROVED** [review]

**Pull request merged**